server/controllers/auth.js

```diff
@@ -121,41 +122,75 @@ exports.loadUser = asyncHandler(async (req, res, next) => {
 exports.logoutUser = asyncHandler(async (req, res, next) => {
   res.clearCookie("token");
 
-  res.send("You have successfully logged out");
+  res.status(200).json({
+    success: {
+      message: "You have successfully logged out",
+    },
+  });
 });
 
 // @route GET /auth/demo-user
 // @desc Load demo user
 // @access Private
 exports.loadDemoUser = asyncHandler(async (req, res, next) => {
-  let demoUser
-  const demoUserExists = await User.findOne({email: "demoUser@email.com"})
+  let demoUser;
+  const demoUserExists = await User.findOne({ email: "demoUser@email.com" });
 
   if (!demoUserExists) {
     demoUser = await User.create({
       email: "demoUser@email.com",
       username: "demoUser",
-      password: "demoUser123"
-    })
+      password: "demoUser123",
+    });
   } else {
-    demoUser = demoUserExists
+    demoUser = demoUserExists;
   }
 
   const token = generateToken(demoUser._id);
   const secondsInWeek = 604800;
 
   res.cookie("token", token, {
     httpOnly: true,
-    maxAge: secondsInWeek * 1000
+    maxAge: secondsInWeek * 1000,
   });
 
   res.status(200).json({
     success: {
       user: {
         id: demoUser._id,
         username: demoUser.username,
-        email: demoUser.email
-      }
-    }
+        email: demoUser.email,
+      },
+    },
+  });
+});
+
+// @route PATCH /auth/update
+// @desc update a user's password
+// @access Private
+exports.updatePassword = asyncHandler(async (req, res, next) => {
+  const { password, oldPassword } = req.body;
+
+  const user = await User.findById(req.user.id).select("+password");
+
+  if (!user) {
+    res.status(400);
+    throw new Error("The user with this ID doesn't exist");
+  }
```

Comment:
I don't fully understand your argument, however from the current `protect` implementation, the middleware only extracts the user info from the token without checking if the user exists in the database. 
I understand that is probably not a standard approach however that is what runs currently.

Except we update the `protect` middleware I think it might be safe to check if the user exists. 

Thinking about it now I think I'd just move the check to the protect middleware and free all the other controllers who use the protect from having to do the same check.

Reply:
Hi @OladapoAjala, sorry for the delay in my response - I think it might still be worth clarifying the idea here.

> Thinking about it now I think I'd just move the check to the protect middleware and free all the other controllers who use the protect from having to do the same check.

Let me start by saying I completely agree that this would be the right way to go about it! 👍🏼 

The `protect` middleware verifies a token that has been given to a user upon successful authentication (e.g. successful registration or login). For a successful authentication to happen, the user **must** have been in the DB at some point in time. 

Considering that, the only possible scenario where a user will have a valid JWT associated with it but it isn't a valid document in our DB anymore, is if it's been deleted somewhere along the way. Now by design, this scenario can be prevented in many ways, one of which is exactly what you suggested - let `protect` query the DB to see if user exists. However, we can usually avoid the extra roundtrip to the DB by designing a user deletion process that avoids this scenario alltogether. And this was my main point - in a real world environment more often than not this scenario will be handled by your `auth` processes and therefore you're free to assume there will never be an invalid user id passed to your handlers when designing your other controllers. 

Hope that clarifies things a little bit 👍🏼